### DISCUSSION TITLE
Init rook-ceph-mon "Permission denied" error

### DISCUSSION THREAD

**Discussion by reisman234** [rook/rook#12225] [Q&A] [answered]
Hello,

I have a cluster with 4 nodes and want to deploy rook-ceph. This is my first time setting up a cluster with the helm charts and followed this [instructions](https://rook.io/docs/rook/v1.11/Helm-Charts/helm-charts/)

I first deploy the operator without any changes on the values.yaml, and after that the cluster itself with the following overwrite-values.yaml

```yaml
# Installs a debugging toolbox deployment
toolbox:
  # -- Enable Ceph debugging pod deployment. See [toolbox](../Troubleshooting/ceph-toolbox.md)
  enabled: true


cephClusterSpec:

  storage: # cluster level storage configuration and selection
    useAllNodes: false
    useAllDevices: false
    # deviceFilter:
    # config:
    #   crushRoot: "custom-root" # specify a non-default root label for the CRUSH map
    #   metadataDevice: "md0" # specify a non-rotational storage so ceph-volume will use it as block db device of bluestore.
    #   databaseSizeMB: "1024" # uncomment if the disks are smaller than 100 GB
    #   journalSizeMB: "1024"  # uncomment if the disks are 20 GB or smaller
    #   osdsPerDevice: "1" # this value can be overridden at the node or device level
    #   encryptedDevice: "true" # the default value for this option is "false"
    # # Individual nodes and their config can be specified as well, but 'useAllNodes' above must be set to false. Then, only the named
    # # nodes below will be used as storage resources. Each node's 'name' field should match their 'kubernetes.io/hostname' label.
    nodes:
      - name: "node-2"
        devices: # specific devices to use for storage can be specified for each node
          - name: "sda"
      - name: "node-3"
        devices:
          - name: "sda"
      - name: "node-4"
        devices:
          - name: "sdb"
      - name: "node-5"
        devices:
          - name: "sdb"
```

After the first pods come to live, some of them went very fast to an error and crashloopbackop state. The logs of one of the failed mon pod showed a permission denied error.

```shell
$ kubectl logs -n rook-ceph rook-ceph-mon-e-5c675f4d4-bjdmf chown-container-data-dir 
chown: cannot access '/var/log/ceph/ceph-volume.log': Permission denied
chown: cannot access '/var/log/ceph/ceph-osd.1.log': Permission denied
failed to change ownership of '/var/log/ceph/ceph-volume.log' to ceph:ceph
chown: changing ownership of '/var/log/ceph': Permission denied
failed to change ownership of '/var/log/ceph/ceph-osd.1.log' to ceph:ceph
failed to change ownership of '/var/log/ceph' from ceph:ceph to ceph:ceph
chown: changing ownership of '/var/lib/ceph/crash': Permission denied
failed to change ownership of '/var/lib/ceph/crash' from ceph:ceph to ceph:ceph
failed to change ownership of '/run/ceph/ceph-osd.1.asok' to ceph:ceph
chown: cannot access '/run/ceph/ceph-osd.1.asok': Permission denied
chown: changing ownership of '/run/ceph': Permission denied
chown: cannot read directory '/var/lib/ceph/mon/ceph-e': Permission denied
failed to change ownership of '/run/ceph' from ceph:ceph to ceph:ceph
failed to change ownership of '/var/lib/ceph/mon/ceph-e' to ceph:ceph
```

**Comment by reisman234** [comment]
After this first failed deployment, I uninstalled the cluster again with helm, But this only removed the release in helm and did not delete any resources in Kubernetes.The pods are still running and trying to come up.

**Reply by travisn** [reply, marked as answer]
Which environment are you running in? It seems the pods need more privileges. You may need to set `hostpathRequiresPrivileged: true` in the `rook-ceph` helm chart.

If you're seeing the resources still exist after removing the charts to attempt a reinstall, see the [Cleanup guide](https://rook.io/docs/rook/latest/Getting-Started/ceph-teardown/) to tear down the rook cluster.

**Reply by reisman234** [reply]
Thanks for the response!

The cluster is set up as a bare metal rke2 Kubernetes distro.
After I saw this error, I did change this setting and updated the operator, but I'm unsure if this would take into account on a running cluster. 
   
 I also looked again in the documentation of ceph and saw that it does not support centos7 anymore, is this also true for rook?

Thanks for the Cleanup guide hint, it helped to tear down the last pieces! 
> Which environment are you running in? It seems the pods need more privileges. You may need to set `hostpathRequiresPrivileged: true` in the `rook-ceph` helm chart.
> 
> If you're seeing the resources still exist after removing the charts to attempt a reinstall, see the [Cleanup guide](https://rook.io/docs/rook/latest/Getting-Started/ceph-teardown/) to tear down the rook cluster.

**Reply by travisn** [reply]
> After I saw this error, I did change this setting and updated the operator, but I'm unsure if this would take into account on a running cluster.

The operator should update the mons with the new privileges, instead of needing to tear down. But if you already tore down, looks like you're in a better place now. 

> I also looked again in the documentation of ceph and saw that it does not support centos7 anymore, is this also true for rook?

Since rook uses containerized ceph it may be fine, but I'm not sure of the ceph limitation that may apply there.

**Reply by reisman234** [reply]
> The operator should update the mons with the new privileges, instead of needing to tear down. But if you already tore down, looks like you're in a better place now.

Yes, I'm now in a better place :smile:. After a new setup with the `hostpathRequiresPrivileged: true` the cluster is now working smoothly. Thanks !

**Comment by ChrisAtkinson78** [comment]
I'm attempting to run rook-ceph on SLES 15, I'm seeing this CHOWN issue when the operator deploys mons:

SLES 15 SP4 / rook/ceph:v1.14.8 / installing with HELM 

Logs(rook-ceph/rook-ceph-mon-a-867df8f6fd-2m8mp:chown-container-data-dir)[tail] ─

chown: changing ownership of '/var/log/ceph': Operation not permitted                                                                                                                                 
│ failed to change ownership of '/var/log/ceph' from root:root to ceph:ceph                                                                                                                             
│ chown: cannot access '/var/lib/ceph/crash': Permission denied                                                                                                                                         
│ failed to change ownership of '/var/lib/ceph/crash' to ceph:ceph                                                                                                                                      
│ chown: changing ownership of '/run/ceph': Operation not permitted                                                                                                                                     
│ failed to change ownership of '/run/ceph' from root:root to ceph:ceph                                                                                                                                 
│ chown: cannot access '/var/lib/ceph/mon/ceph-a': Permission denied                                                                                                                                    
│ failed to change ownership of '/var/lib/ceph/mon/ceph-a' to ceph:ceph 

I've tried the hostpathRequiresPrivileged:  true 
Here is the snippet from the running operator:

Environment:                                                                                                                                                                                      │
│       ROOK_CURRENT_NAMESPACE_ONLY:        false                                                                                                                                                       
│       ROOK_HOSTPATH_REQUIRES_PRIVILEGED:  true                                                                                                                                                        
│       ROOK_DISABLE_DEVICE_HOTPLUG:        false                                                                                                                                                       
│       ROOK_DISCOVER_DEVICES_INTERVAL:     60m                                                                                                                                                         
│       NODE_NAME:                           (v1:spec.nodeName)                                                                                                                                         
│       POD_NAME:                           rook-ceph-operator-7499d47895-c7794 (v1:metadata.name)                                                                                                      
│       POD_NAMESPACE:                      rook-ceph (v1:metadata.namespace)   

Any ideas?  I've attempted to modify the continaerSecurityContext:
containerSecurityContext:
  runAsNonRoot: false
  runAsUser: 0
  runAsGroup: 0
  capabilities:
    drop: ["ALL"]

**Reply by travisn** [reply]
I would expect the setting `hostpathRequiresPrivileged: true` to resolve this issue. Did you try deploying fresh after changing this setting? Perhaps the mon pods aren't yet picking up this setting. Does the security context for the mons show that they are privileged now?

**Reply by ChrisAtkinson78** [reply]
I resolved this issue.  I was using helm to install the operator and then kubectl to apply a cluster.yaml.  I didn't realize this wouldn't include defaults that helm typically provides.  So removed and reinstalled with helm for both operator and cluster and it came right up.  Thanks for the followup!
